core/common/scale.cpp

```diff
@@ -0,0 +1,506 @@
+      }
+
+      EncodeIntegerResult encodeInt8(int8_t value) noexcept {
+        return encodeUint8(util::makeSignReinterpreter(value).asUnsigned());
+      }
+
+      EncodeIntegerResult encodeUint16(uint16_t value) noexcept {
+        if (value < EncodingCategoryLimits::kLessThan16384) {
+          uint16_t v = value;
+          v <<= 2;  // v *= 4
+          ++v;      // set 0b01 flag
+          auto minor_byte = static_cast<uint8_t>(v & 0xFF);
+          v >>= 8;
+          auto major_byte = static_cast<uint8_t>(v & 0xFF);
+          return expected::Value{ByteArray{minor_byte, major_byte}};
+        }
+
+        return encodeUint32(static_cast<uint32_t>(value));
+      }
+
+      EncodeIntegerResult encodeInt16(int16_t value) noexcept {
+        return encodeUint16(util::makeSignReinterpreter(value).asUnsigned());
+      }
+
+      EncodeIntegerResult encodeUint32(uint32_t value) noexcept {
+        if (value < EncodingCategoryLimits::kLessThanTwoDegreeThirty) {
+          uint32_t value_le = impl::encodeLittleEndian((value << 2) + 2);
+
+          impl::AsArray<uint32_t> arr{value_le};
+
+          return expected::Value{ByteArray(arr.begin(), arr.end())};
+        }
+
+        return encodeUint64(static_cast<uint64_t>(value));
+      }
+
+      EncodeIntegerResult encodeInt32(int32_t value) noexcept {
+        return encodeUint32(util::makeSignReinterpreter(value).asUnsigned());
+      }
+
+      EncodeIntegerResult encodeUint64(uint64_t value) noexcept {
+        return encodeBigInteger(BigInteger{value});
+      }
+
+      EncodeIntegerResult encodeInt64(int64_t value) noexcept {
+        if (value < 0)
+          return expected::Error{EncodeBigIntegerError::kBigNegativeValue};
+
+        return encodeUint64(static_cast<uint64_t>(value));
+      }
+
+      EncodeIntegerResult encodeBigInteger(const BigInteger &value) noexcept {
+        // cannot encode negative numbers
+        if (value < 0)
+          return expected::Error{EncodeBigIntegerError::kBigNegativeValue};
+
+        // calculate number of bytes required
+        auto countBytes = [](BigInteger v) -> size_t {
+          if (v == 0)
+            return 1;
+
+          size_t counter = 0;
+          while (v > 0) {
+            ++counter;
+            v /= 256;
+          }
+
+          return counter;
+        };
+
+        // number of bytes required to represent value
+        size_t bigIntLength = countBytes(value);
+
+        // number of bytes to scale-encode value
+        // 1 byte is reserved for header
+        size_t requiredLength = 1 + bigIntLength;
+
+        // not sure return error or just encode as is
+        // bigIntLength - 4 is encoded in value
+        if (bigIntLength < 4)
+          bigIntLength = 4;
+
+        if (bigIntLength > 67) {
+          return expected::Error{EncodeBigIntegerError::kValueIsTooBig};
+        }
+
+        ByteArray result;
+        result.reserve(requiredLength);
+        uint8_t header = (bigIntLength - 4) * 4 + 3;
```

Comment:
would be nice to add a comment about constants here

Reply:
// the value encoded in 6 major bits of the first byte is
        // number of bytes for encoding big integer minus 4
        // big integers require at least 4 bytes
        // so subtracting 4 from the value can increase maximum number of bytes
        // encoded in the first byte by 4, and this max value is 67
        uint8_t header =
            (bigIntLength - 4)
            * 4 // move left by 2 bits
            + 3; // 3 == 0b11 means fourth case of encoding

Ok?

Reply:
D: don't split statements is such a way, imo that looks ugly

I mean mostly explain header variable here in "high-level", if possible. Something like "Header is simply a length of big integer, starting from 4 bytes" is enough, i think. Also note, that `bigIntLength` assumed to be in range 1..(2^6-1) might be also handy

Reply:
`/* The value stored in 6 major bits of header is used
         * to encode number of bytes for storing big integer.
         * Value formed by 6 bits varies from 0 to 63 == 2^6 - 1,
         * However big integer byte count starts from 4,
         * so to store this number we should decrease this value by 4.
         * And the range of bytes number for storing big integer becomes 4 .. 67.
         * To form resulting header we need to move that bits
         * representing bytes count to the left by 2 positions
         * by means of multiplying by 4.
         * Minor 2 bits store encoding option, in our case it is 0b11 == 3
         * We just add 3 to the result of operations above
         */
        uint8_t header = (bigIntLength - 4) * 4 + 3;`
Is this ok?